### ISSUE TITLE
Assert failure in user mode Multi-threaded stress test

### ISSUE THREAD

**Issue by dv-msft** [microsoft/ebpf-for-windows#2323]
### Describe the bug

_NOTE: This needs changes from PR #2319 as it fixes an upstream concurrency issue.  That fix is required for the test to get to this point._

Running the user mode multi-thread stress test application causes an assert:

```
************* Preparing the environment for Debugger Extensions Gallery repositories **************
   ExtensionRepository : Implicit
   UseExperimentalFeatureForNugetShare : false
   AllowNugetExeUpdate : false

   - Configuring repositories
      ----> Repository : Internal, Enabled: true
      ----> Repository : LocalInstalled, Enabled: true
      ----> Repository : UserExtensions, Enabled: true


************* Waiting for Debugger Extensions Gallery to Initialize **************
.
   ----> Repository : LocalInstalled, Enabled: true, Packages count: 44
   ----> Repository : UserExtensions, Enabled: true, Packages count: 0
   ----> Repository : Internal, Enabled: true, Packages count: 41

Microsoft (R) Windows Debugger Version 10.0.25324.1001 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

CommandLine: .\ebpf_stress_tests_um -jtt=32 -jtd=300

************* Path validation summary **************
Response                         Time (ms)     Location
Deferred                                       srv*
Deferred                                       cache*C:\SymCache
Deferred                                       srv*http://symcache
Deferred                                       srv*http://symweb
Symbol search path is: srv*;cache*C:\SymCache;srv*http://symcache;srv*http://symweb
Executable search path is: 
ModLoad: 00007ff7`00060000 00007ff7`006b8000   ebpf_stress_tests_um.exe
ModLoad: 00007ffc`86a30000 00007ffc`86c48000   ntdll.dll
ModLoad: 00007ffc`852f0000 00007ffc`853b3000   C:\WINDOWS\System32\KERNEL32.DLL
ModLoad: 00007ffc`84190000 00007ffc`844e0000   C:\WINDOWS\System32\KERNELBASE.dll
ModLoad: 00007ffc`812d0000 00007ffc`8136b000   C:\WINDOWS\SYSTEM32\apphelp.dll
ModLoad: 00007ffc`854a0000 00007ffc`855c7000   C:\WINDOWS\System32\RPCRT4.dll
ModLoad: 00007ffc`035b0000 00007ffc`03692000   D:\wrk\main-ebpf-for-windows\x64\Debug\MSVCP140D.dll
ModLoad: 00007ffc`6d010000 00007ffc`6d241000   C:\WINDOWS\SYSTEM32\dbghelp.dll
ModLoad: 00007ffc`81750000 00007ffc`81835000   C:\WINDOWS\SYSTEM32\wer.dll
ModLoad: 00007ffc`838b0000 00007ffc`838d8000   C:\WINDOWS\SYSTEM32\bcrypt.dll
ModLoad: 00007ffc`85b40000 00007ffc`85cd2000   C:\WINDOWS\System32\ole32.dll
ModLoad: 00007ffc`84580000 00007ffc`84690000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007ffc`4de60000 00007ffc`4de8b000   D:\wrk\main-ebpf-for-windows\x64\Debug\VCRUNTIME140D.dll
ModLoad: 00007ffc`84690000 00007ffc`84734000   C:\WINDOWS\System32\msvcp_win.dll
ModLoad: 00000219`f8970000 00000219`f8a80000   C:\WINDOWS\System32\ucrtbase.dll
ModLoad: 00007ffc`849a0000 00007ffc`849c9000   C:\WINDOWS\System32\GDI32.dll
ModLoad: 00007ffc`5fdf0000 00007ffc`5fdff000   D:\wrk\main-ebpf-for-windows\x64\Debug\VCRUNTIME140_1D.dll
ModLoad: 00007ffc`85680000 00007ffc`85a0a000   C:\WINDOWS\System32\combase.dll
ModLoad: 00007ffc`84550000 00007ffc`84576000   C:\WINDOWS\System32\win32u.dll
ModLoad: 00007ffb`eca50000 00007ffb`ecc71000   D:\wrk\main-ebpf-for-windows\x64\Debug\ucrtbased.dll
ModLoad: 00007ffc`853c0000 00007ffc`85491000   C:\WINDOWS\System32\OLEAUT32.dll
ModLoad: 00000219`f8970000 00000219`f8b91000   D:\wrk\main-ebpf-for-windows\x64\Debug\ucrtbased.dll
ModLoad: 00007ffc`84070000 00007ffc`8418e000   C:\WINDOWS\System32\gdi32full.dll
ModLoad: 00007ffc`85e10000 00007ffc`85fc2000   C:\WINDOWS\System32\USER32.dll
(2a40.8648): Break instruction exception - code 80000003 (first chance)
ntdll!LdrpDoDebuggerBreak+0x30:
00007ffc`86b53c34 cc              int     3
0:000> g
ModLoad: 00007ffc`84aa0000 00007ffc`84ad0000   C:\WINDOWS\System32\IMM32.DLL
ModLoad: 00007ffc`85fd0000 00007ffc`8607f000   C:\WINDOWS\System32\advapi32.dll
ModLoad: 00007ffc`855d0000 00007ffc`85676000   C:\WINDOWS\System32\msvcrt.dll
ModLoad: 00007ffc`85d40000 00007ffc`85de3000   C:\WINDOWS\System32\sechost.dll
ModLoad: 00007ffc`83750000 00007ffc`8375c000   C:\WINDOWS\SYSTEM32\CRYPTBASE.DLL
ModLoad: 00007ffc`84740000 00007ffc`847c7000   C:\WINDOWS\System32\bcryptPrimitives.dll
ModLoad: 00007ffc`82f10000 00007ffc`82f28000   C:\WINDOWS\SYSTEM32\kernel.appcore.dll
Debug Error!

Program: D:\wrk\main-ebpf-for-windows\x64\Debug\ebpf_stress_tests_um.exe

abort() has been called

(Press Retry to debug the application)
ModLoad: 00007ffc`70ce0000 00007ffc`70d8b000   C:\WINDOWS\SYSTEM32\TextShaping.dll
ModLoad: 00007ffc`813d0000 00007ffc`81479000   C:\WINDOWS\system32\uxtheme.dll
ModLoad: 00007ffc`85a10000 00007ffc`85b36000   C:\WINDOWS\System32\MSCTF.dll
ModLoad: 00007ffc`6fd00000 00007ffc`6fe38000   C:\WINDOWS\SYSTEM32\textinputframework.dll
ModLoad: 00007ffc`81010000 00007ffc`8113c000   C:\WINDOWS\SYSTEM32\CoreMessaging.dll
ModLoad: 00007ffc`7e290000 00007ffc`7e5bd000   C:\WINDOWS\SYSTEM32\CoreUIComponents.dll
ModLoad: 00007ffc`7dfb0000 00007ffc`7e114000   C:\WINDOWS\SYSTEM32\wintypes.dll
(2a40.7ef4): Break instruction exception - code 80000003 (first chance)
ucrtbased!issue_debug_notification+0x45:
00007ffb`ecae2fb5 cc              int     3
0:070> k
 # Child-SP          RetAddr               Call Site
00 000000da`6f0fd1b0 00007ffb`ecae3153     ucrtbased!issue_debug_notification+0x45 [minkernel\crts\ucrt\src\appcrt\internal\report_runtime_error.cpp @ 28] 
01 000000da`6f0fd200 00007ffb`ecafae2d     ucrtbased!__acrt_report_runtime_error+0x13 [minkernel\crts\ucrt\src\appcrt\internal\report_runtime_error.cpp @ 154] 
02 000000da`6f0fd260 00007ffb`ecb01345     ucrtbased!abort+0x1d [minkernel\crts\ucrt\src\appcrt\startup\abort.cpp @ 61] 
03 000000da`6f0fd2a0 00007ffb`ecb00bd7     ucrtbased!common_assert_to_stderr_direct+0xe5 [minkernel\crts\ucrt\src\appcrt\startup\assert.cpp @ 161] 
04 000000da`6f0fd780 00007ffb`ecafebc8     ucrtbased!common_assert_to_stderr<wchar_t>+0x27 [minkernel\crts\ucrt\src\appcrt\startup\assert.cpp @ 179] 
05 000000da`6f0fd7e0 00007ffb`ecb018af     ucrtbased!common_assert<wchar_t>+0x68 [minkernel\crts\ucrt\src\appcrt\startup\assert.cpp @ 420] 
06 000000da`6f0fd820 00007ff7`0029cd56     ucrtbased!_wassert+0x2f [minkernel\crts\ucrt\src\appcrt\startup\assert.cpp @ 444] 
07 000000da`6f0fd850 00007ff7`002a4e8a     ebpf_stress_tests_um!_nmr::remove<std::map<void *,_nmr::client_registration,std::less<void *>,std::allocator<std::pair<void * const,_nmr::client_registration> > > >+0x1d6 [D:\wrk\main-ebpf-for-windows\libs\platform\user\nmr_impl.cpp @ 312] 
08 000000da`6f0fd9a0 00007ff7`0028b82d     ebpf_stress_tests_um!_nmr::wait_for_deregister_client+0x2a [D:\wrk\main-ebpf-for-windows\libs\platform\user\nmr_impl.cpp @ 73] 
09 000000da`6f0fd9d0 00007ff7`00273e55     ebpf_stress_tests_um!NmrWaitForClientDeregisterComplete+0x2d [D:\wrk\main-ebpf-for-windows\libs\platform\user\nmr_um.cpp @ 100] 
0a 000000da`6f0fda20 00007ff7`00273870     ebpf_stress_tests_um!ebpf_extension_unload+0x125 [D:\wrk\main-ebpf-for-windows\libs\platform\ebpf_extension.c @ 338] 
0b 000000da`6f0fdc30 00007ff7`00228ada     ebpf_stress_tests_um!ebpf_extension_load+0x5d0 [D:\wrk\main-ebpf-for-windows\libs\platform\ebpf_extension.c @ 285] 
0c 000000da`6f0fdf50 00007ff7`0021a7c0     ebpf_stress_tests_um!ebpf_link_attach_program+0x29a [D:\wrk\main-ebpf-for-windows\libs\execution_context\ebpf_link.c @ 241] 
0d 000000da`6f0fe1c0 00007ff7`0021d24f     ebpf_stress_tests_um!_ebpf_core_protocol_link_program+0x220 [D:\wrk\main-ebpf-for-windows\libs\execution_context\ebpf_core.c @ 1207] 
0e 000000da`6f0fe3c0 00007ff7`000862f5     ebpf_stress_tests_um!ebpf_core_invoke_protocol_handler+0x2af [D:\wrk\main-ebpf-for-windows\libs\execution_context\ebpf_core.c @ 2430] 
0f 000000da`6f0fe470 00007ff7`0007cd81     ebpf_stress_tests_um!GlueDeviceIoControl+0x2b5 [D:\wrk\main-ebpf-for-windows\tests\end_to_end\test_helper.cpp @ 453] 
10 000000da`6f0fe5d0 00007ff7`00073b93     ebpf_stress_tests_um!std::invoke<bool (__cdecl*&)(void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *),void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *>+0xa1 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\type_traits @ 1589] 
11 000000da`6f0fe640 00007ff7`000885fa     ebpf_stress_tests_um!std::_Invoker_ret<int>::_Call<bool (__cdecl*&)(void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *),void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *>+0x73 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\functional @ 672] 
12 000000da`6f0fe6a0 00007ff7`0006da28     ebpf_stress_tests_um!std::_Func_impl_no_alloc<bool (__cdecl*)(void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *),int,void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *>::_Do_call+0x7a [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\functional @ 834] 
13 000000da`6f0fe700 00007ff7`0006e043     ebpf_stress_tests_um!std::_Func_class<int,void *,unsigned long,void *,unsigned long,void *,unsigned long,unsigned long *,_OVERLAPPED *>::operator()+0xc8 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\functional @ 875] 
14 000000da`6f0fe780 00007ff7`001ae9bf     ebpf_stress_tests_um!Platform::DeviceIoControl+0x73 [D:\wrk\main-ebpf-for-windows\libs\thunk\mock\mock.cpp @ 32] 
15 000000da`6f0fe7f0 00007ff7`001c0f58     ebpf_stress_tests_um!invoke_ioctl<std::vector<unsigned char,std::allocator<unsigned char> >,_ebpf_operation_link_program_reply>+0xcf [D:\wrk\main-ebpf-for-windows\libs\api_common\device_helper.hpp @ 99] 
16 000000da`6f0fe8a0 00007ff7`001d22e4     ebpf_stress_tests_um!_link_ebpf_program+0x3e8 [D:\wrk\main-ebpf-for-windows\libs\api\ebpf_api.cpp @ 1058] 
17 000000da`6f0fea30 00007ff7`000d879d     ebpf_stress_tests_um!ebpf_program_attach_by_fd+0x554 [D:\wrk\main-ebpf-for-windows\libs\api\ebpf_api.cpp @ 1202] 
18 000000da`6f0feb90 00007ff7`000d67db     ebpf_stress_tests_um!_hook_helper::attach_link+0x3d [D:\wrk\main-ebpf-for-windows\tests\end_to_end\helpers.h @ 115] 
19 000000da`6f0febd0 00007ff7`000bdaf6     ebpf_stress_tests_um!_droppacket_stress_thread_function+0x7bb [D:\wrk\main-ebpf-for-windows\tests\stress\ebpf_stress_tests.cpp @ 250] 
1a 000000da`6f0ffc80 00007ff7`000ad53f     ebpf_stress_tests_um!std::invoke<void (__cdecl*)(stress_test_thread_context const &),stress_test_thread_context>+0x36 [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\type_traits @ 1589] 
1b 000000da`6f0ffcc0 00007ffb`ecb03010     ebpf_stress_tests_um!std::thread::_Invoke<std::tuple<void (__cdecl*)(stress_test_thread_context const &),stress_test_thread_context>,0,1>+0x8f [C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.35.32215\include\thread @ 56] 
1c 000000da`6f0ffd40 00007ffc`852f3db1     ucrtbased!thread_start<unsigned int (__cdecl*)(void *),1>+0xb0 [minkernel\crts\ucrt\src\appcrt\startup\thread.cpp @ 97] 
1d 000000da`6f0ffda0 00007ffc`86a332a1     KERNEL32!BaseThreadInitThunk+0x21 [clientcore\base\win32\client\thread.c @ 77] 
1e 000000da`6f0ffdd0 00000000`00000000     ntdll!RtlUserThreadStart+0x21 [minkernel\ldr\rtlstrt.c @ 1198] 

```

### Steps taken to reproduce bug

Run the user mode multi-thread stress test application with the following command-line:

```
.\ebpf_stress_tests_um -jtt=32 -jtd=300
```

### Expected behavior

The test should pass without errors.

### Actual outcome

The test fails in the first minute or so with the following console output:

```
Randomness seeded to: 885189179
jit test programs:
        droppacket
jit test threads per program: 32
jit test duration: 300
jit test verbose output: false
One time initialization complete
spawning stress test threads for droppacket.o...
waiting on 32 threads...
Assertion failed: it->second.binding_count == 0, file D:\wrk\main-ebpf-for-windows\libs\platform\user\nmr_impl.cpp, line 312
```

**Comment by dv-msft** [comment]
A lot of fixes have gone in since this was first reported and this issue has not been repro'd in repeated testing in local testing nor has it been reported by CI/CD.  Closing as 'No Repro'.
